README.md
<h2 align="center">this-is-ok</h2>

  <p align="center">
    An ergonomic way to work with fallible operations in TypeScript.
    <br />
    <br />
    <a href="https://github.com/mieszkosabo/this-is-ok#api-reference-">API Reference</a>
    ·
    <a href="https://github.com/mieszkosabo/this-is-ok/issues/new?assignees=&labels=bug&projects=&template=bug_report.md&title=">Report Bug</a>
    ·
    <a href="https://github.com/mieszkosabo/this-is-ok/issues/new">Request Feature</a>
    ·
    <a href="https://github.com/mieszkosabo/this-is-ok#road-map-">Road map</a>
  </p>

---

## Why this library? <!-- omit in toc -->

- ✨ Fully type-safe, and ergonomic API that uses popular conventions.
- 🎉 Comes with a clean and easy to use way to simulate the do-notation.
- ⚡️ Tree-shakable, works with esm, cjs, and doesn't force you to use `nodenext`/`node16` module resolution.

## Getting started <!-- omit in toc -->

### Installation <!-- omit in toc -->

```bash
npm install this-is-ok

yarn add this-is-ok

pnpm add this-is-ok

```

### Basic usage <!-- omit in toc -->

```ts
// an example of adapting a standard library function to use Option
const safeParseInt = (value: string): Option<number> => {
  const parsed = parseInt(value);
  if (isNaN(parsed)) {
    return none;
  }
  return some(parsed);
};

// `of` from the option module creates an Option from a nullable value
of(localStorage.getItem("counter"))
  // map the inner value to another Option
  .flatMap((d) => safeParseInt(d))
  // map the inner value or use a default value if it's none
  .mapOr(0, (d) => d + 1)
  // perform any side effect with the value
  .tap((d) => localStorage.setItem("counter", d.toString()));
```

### do notation <!-- omit in toc -->

Do notation lets us write the "happy path" of a function without having to worry about the error handling.
We access the inner values of options by calling `bind` on them, and if any point, inside the `do` block, `bind` is called on `none`
then we short-circuit the execution and none is returned.

```ts
some(1)
  .do((value) => {
    const a = some(2).bind();
    const b = some("hello").bind();
    return some(b.repeat(value + a));
  })
  .tap(console, log); // "hellohellohello"

some(1).do((value) => {
  const a = value; // 1

  const b = some(2).bind(); // 2

  const c = (none as Option<number>).bind();
  // the execution stops here and none is returned
  // so the following lines are not executed

  const d = some(3).bind();
  return some(a + b + c + d);
}).isNone; // true
```

# Api Reference <!-- omit in toc -->

- [Option](#option)
  - [Creation](#creation)
    - [`of`](#of)
    - [`from`](#from)
  - [Methods](#methods)
    - [`isSome: boolean`](#issome-boolean)
    - [`isSomeAnd: (predicate: (value: T) => boolean) => boolean`](#issomeand-predicate-value-t--boolean--boolean)
    - [`isNone: boolean`](#isnone-boolean)
    - [`expect: (message: string) => T`](#expect-message-string--t)
    - [`unwrap: () => T`](#unwrap---t)
    - [`unwrapOr: (defaultValue: T) => T`](#unwrapor-defaultvalue-t--t)
    - [`unwrapOrElse: (defaultValueFun: () => T) => T`](#unwraporelse-defaultvaluefun---t--t)
    - [`map: <U>(f: (value: T) => U) => Option<U>`](#map-uf-value-t--u--optionu)
    - [`mapOr: <U>(defaultValue: U, f: (value: T) => U) => U`](#mapor-udefaultvalue-u-f-value-t--u--u)
    - [`mapOrElse: <U>(defaultValueFun: () => U, f: (arg: T) => U) => U`](#maporelse-udefaultvaluefun---u-f-arg-t--u--u)
    - [`okOr: <E>(err: E) => Result<T, E>`](#okor-eerr-e--resultt-e)
    - [`okOrElse: <E>(f: () => E) => Result<T, E>`](#okorelse-ef---e--resultt-e)
    - [`and: <U>(b: Option<U>) => Option<U>`](#and-ub-optionu--optionu)
    - [`or: (b: Option<T>) => Option<T>`](#or-b-optiont--optiont)
    - [`orElse: (f: () => Option<T>) => Option<T>`](#orelse-f---optiont--optiont)
    - [`flatMap: <U>(f: (value: T) => Option<U>) => Option<U>`](#flatmap-uf-value-t--optionu--optionu)
    - [`andThen: <U>(f: (value: T) => Option<U>) => Option<U>`](#andthen-uf-value-t--optionu--optionu)
    - [`tap: <F extends void | Promise<void>>(f: (value: T) => F) => F`](#tap-f-extends-void--promisevoidf-value-t--f--f)
    - [`filter: (predicate: (arg: T) => boolean) => Option<T>`](#filter-predicate-arg-t--boolean--optiont)
    - [`match: <U>(pattern: { some: (value: T) => U; none: () => U }) => U`](#match-upattern--some-value-t--u-none---u---u)
    - [`do: <U>(f: (value: T) => Option<U>) => Option<U>`](#do-uf-value-t--optionu--optionu)
- [Result](#result)
  - [Creation](#creation-1)
    - [`of = <T, E>(value: T, error: E): Result<T, E>`](#of--t-evalue-t-error-e-resultt-e)
    - [`from = <T, E>(fn: () => T, error: E): Result<T, E>`](#from--t-efn---t-error-e-resultt-e)
    - [`<T, E extends Error>(fn: () => T): Result<T, E>`](#t-e-extends-errorfn---t-resultt-e)
  - [Methods](#methods-1)
    - [`isOk: boolean`](#isok-boolean)
    - [`isOkAnd: (predicate: (value: T) => boolean) => boolean`](#isokand-predicate-value-t--boolean--boolean)
    - [`isErr: boolean`](#iserr-boolean)
    - [`isErrAnd: (predicate: (value: E) => boolean) => boolean`](#iserrand-predicate-value-e--boolean--boolean)
    - [`ok: () => Option<T>`](#ok---optiont)
    - [`err: () => Option<T>`](#err---optiont)
    - [`expect: (message: string) => T`](#expect-message-string--t-1)
    - [`unwrap: () => T`](#unwrap---t-1)
    - [`expectErr: (message: string) => T`](#expecterr-message-string--t)
    - [`unwrapErr: () => E`](#unwraperr---e)
    - [`unwrapOr: (defaultValue: T) => T`](#unwrapor-defaultvalue-t--t-1)
    - [`unwrapOrElse: (defaultValueFun: () => T) => T`](#unwraporelse-defaultvaluefun---t--t-1)
    - [`map: <U>(f: (value: T) => U) => Result<U, E>`](#map-uf-value-t--u--resultu-e)
    - [`mapOr: <U>(defaultValue: U, f: (value: T) => U) => U`](#mapor-udefaultvalue-u-f-value-t--u--u-1)
    - [`mapOrElse: <U>(defaultValueFun: () => U, f: (arg: T) => U) => U`](#maporelse-udefaultvaluefun---u-f-arg-t--u--u-1)
    - [`mapErr: <U>(f: (value: E) => U) => Result<T, U>`](#maperr-uf-value-e--u--resultt-u)
    - [`and: <U>(b: Result<U, E>) => Result<U, E>`](#and-ub-resultu-e--resultu-e)
    - [`or: <F>(b: Result<T, F>) => Result<T,F>`](#or-fb-resultt-f--resulttf)
    - [`orElse: <F>(f: () => Result<T, F>) => Result<T, F>`](#orelse-ff---resultt-f--resultt-f)
    - [`flatMap: <U>(f: (value: T) => Result<U, E>) => Result<U, E>`](#flatmap-uf-value-t--resultu-e--resultu-e)
    - [`andThen: <U>(f: (value: T) => Result<U, E>) => Result<U, E>`](#andthen-uf-value-t--resultu-e--resultu-e)
    - [`do: <U, F>(f: (value: T) => Result<U, F>) => Result<U, F>`](#do-u-ff-value-t--resultu-f--resultu-f)
    - [`match: <U>(pattern: { ok: (value: T) => U; err: (value: E) => U }) => U`](#match-upattern--ok-value-t--u-err-value-e--u---u)

## Option

### Creation

#### `of`

Converts a nullable value into an option.

Example:

```ts
of(1); // some(1)
of(null); // none
of(undefined); // none
```

#### `from`

Calls a function and converts its result into an option. If the function
is null or undefined or it throws an exception, none is returned.

Example:

```ts
from(() => 1); // some(1)
from(() => null); // none
from(() => {
  throw new Error("oopsie hehe");
}); // none
```

### Methods

#### `isSome: boolean`

Property that is `true` if the option is a `Some` variant and `false` otherwise.

Example:

```ts
some(3).isSome; // true
none.isSome; // false
```

#### `isSomeAnd: (predicate: (value: T) => boolean) => boolean`

Returns true if the option is a Some variant and the value inside
of it matches a predicate.

Example:

```ts
expect(some(42).isSomeAnd((v) => v % 2)).toBe(true);
expect(some(42).isSomeAnd((v) => v > 100)).toBe(true);
```

#### `isNone: boolean`

Property that is `true` if the option is a `None` variant and `false` otherwise.

Example:

```ts
expect(none.isNone).toBe(true);
expect(some(42).isNone).toBe(false);
```

#### `expect: (message: string) => T`

Returns the inner value. Useful for situations where if the
option is a `None` then there is no tap to continue running the program.
The message can be used for debugging purposes.

**Parameters:**

- `message` - The message to throw if the option is a `None` variant.

**Throws:**

Throws the given message if the option is a `None` variant.

**Example:**

```ts
expect(some(42).expect("should be a number")).toBe(42);
expect(none.expect("this will throw")).toThrow();
```

#### `unwrap: () => T`

Returns the inner value or throws an Error if the option is a `None` variant.

**Throws:**

Throws if the option is a `None` variant.

**Example:**

```ts
expect(some(42).unwrap()).toBe(42);
expect(none.unwrap()).toThrow();
```

#### `unwrapOr: (defaultValue: T) => T`

Returns the inner value or the provided default value.

**Example:**

```ts
expect(some(42).unwrapOr(1)).toBe(42);
expect(none.unwrapOr(1)).toBe(1);
```

#### `unwrapOrElse: (defaultValueFun: () => T) => T`

Returns the inner value or computes it from a function.

**Example:**

```ts
expect(some(42).unwrapOrElse(() => 1)).toBe(42);
expect(none.unwrapOrElse(() => 1)).toBe(1);
```

#### `map: <U>(f: (value: T) => U) => Option<U>`

Maps an `Option<T>` to `Option<U>` by applying a function to a contained
value (if Some) or returns None (if None).

**Example:**

```ts
expect(
  some(42)
    .map((x) => x + 1)
    .unwrap()
).toBe(43);
expect(none.map((x) => x + 1).isNone()).toBe(true);
```

#### `mapOr: <U>(defaultValue: U, f: (value: T) => U) => U`

Returns the provided default result (if none),
or applies a function to the contained value (if any).

**Example:**

```ts
expect(
  some(42)
    .mapOr(10, (x) => x + 1)
    .unwrap()
).toBe(43);
expect(none.mapOr(10, (x) => x + 1).unwrap()).toBe(11);
```

#### `mapOrElse: <U>(defaultValueFun: () => U, f: (arg: T) => U) => U`

Computes a default function result (if none),
or applies a different function to the contained value (if any).

**Example:**

```ts
expect(
  some(42)
    .mapOrElse(
      () => 10,
      (x) => x + 1
    )
    .unwrap()
).toBe(43);
expect(
  none
    .mapOrElse(
      () => 10,
      (x) => x + 1
    )
    .unwrap()
).toBe(11);
```

#### `okOr: <E>(err: E) => Result<T, E>`

Given an `error: E` converts `Option<T>` to `Result<T, E>`.

**Example:**

```ts
some(42).okOr("error"); // { variant: "ok", value: 42, ... }
none.okOr("error"); // { variant: "error", error: "error", ... }
```

#### `okOrElse: <E>(f: () => E) => Result<T, E>`

Same as `okOr` but the error is computed from a function.

#### `and: <U>(b: Option<U>) => Option<U>`

Returns None if the option is None, otherwise returns `b`.

**Example:**

```ts
expect(some(42).and(some(1)).unwrap()).toBe(1);
expect(none.and(some(1)).is

None()).toBe(true);
```

#### `or: (b: Option<T>) => Option<T>`

Returns None if the option is None, otherwise returns `b`.

**Example:**

```ts
expect(some(42).or(some(1)).unwrap()).toBe(43);
expect(none.or(some(1)).unwrap()).toBe(1);
```

#### `orElse: (f: () => Option<T>) => Option<T>`

Returns the option if it contains a value,
otherwise calls f and returns the result.

**Example:**

```ts
expect(
  some(42)
    .orElse(() => some(1))
    .unwrap()
).toBe(42);
expect(none.orElse(() => some(1)).unwrap()).toBe(1);
```

#### `flatMap: <U>(f: (value: T) => Option<U>) => Option<U>`

Returns None if the option is None, otherwise calls
f with the wrapped value and returns the result.

**Example:**

```ts
expect(
  some(42)
    .flatMap((x) => some(x + 1))
    .unwrap()
).toBe(43);
expect(
  some(42)
    .flatMap((x) => none)
    .isNone()
).toBe(true);
```

#### `andThen: <U>(f: (value: T) => Option<U>) => Option<U>`

Returns None if the option is None, otherwise calls
f with the wrapped value and returns the result.

**Example:**

```ts
expect(
  some(42)
    .andThen((x) => some(x + 1))
    .unwrap()
).toBe(43);
expect(
  some(42)
    .andThen((x) => none)
    .isNone()
).toBe(true);
```

#### `tap: <F extends void | Promise<void>>(f: (value: T) => F) => F`

Runs the given void function with the inner value if the option is a `Some` variant
or does nothing if the option is a `None` variant.

**Example:**

```ts
expect(
  some(42)
    .tap((x) => some(x + 1))
    .unwrap()
).toBe(43);
expect(
  some(42)
    .tap((x) => none)
    .isNone()
).toBe(true);
```

#### `filter: (predicate: (arg: T) => boolean) => Option<T>`

Returns None if the option is None, otherwise returns `None` if the predicate `predicate` returns `false` when applied to the contained
value, otherwise returns the `Option<T>`.

**Example:**

```ts
expect(
  some(42)
    .filter((x) => x > 10)
    .unwrap()
).toBe(42);
expect(
  some(42)
    .filter((x) => x < 10)
    .isNone()
).toBe(true);
```

#### `match: <U>(pattern: { some: (value: T) => U; none: () => U }) => U`

Allows you to run different functions depending on the variant of the option.

**Example:**

```ts
expect(
  some(3).match({
    some: (d) => d + 1,
    none: () => 0,
  })
).toBe(4);

expect(
  none.match({
    some: (d) => d + 1,
    none: () => 0,
  })
).toBe(0);
```

#### `do: <U>(f: (value: T) => Option<U>) => Option<U>`

Lets you simulate a do-notation known from functional languages
with the Option monad.

**Example:**

```ts
expect(
  some(1)
    .do((value) => {
      const a = value;
      const b = some(2).bind();
      const c = some(3).bind();
      return some(a + b + c);
    })
    .unwrap()
).toBe(6);
```

## Result

### Creation

#### `of = <T, E>(value: T, error: E): Result<T, E>`

Creates a new `Result<T, E>` from a possibly nullable value and an error.

**Example:**

```ts
expect(of(42, "error").unwrap()).toBe(42);
expect(of({}, "error").unwrap()).toEqual({});
expect(of(NaN, "error").unwrap()).toEqual(NaN);
expect(of("", "error").unwrap()).toEqual("");

expect(of(null, "error").isErr).toBe(true);
expect(of(undefined, "error").isErr).toBe(true);
```

#### `from = <T, E>(fn: () => T, error: E): Result<T, E>`

Creates a new `Result<T, E>` from a function that can throw and an error.

**Example:**

```ts
expect(from(() => 42, "error").unwrap()).toBe(42);
expect(from(() => ({}), "error").unwrap()).toEqual({});
expect(from(() => NaN, "error").unwrap()).toEqual(NaN);
expect(from(() => "", "error").unwrap()).toEqual("");

expect(from(() => null, "error").isErr).toBe(true);
expect(from(() => undefined, "error").isErr).toBe(true);

expect(
  from(() => {
    throw "error";
  }, "error").isErr
).toBe(true);
```

#### `<T, E extends Error>(fn: () => T): Result<T, E>`

Creates a new `Result<T, E>` from a function that can throw.
If the function returns any value (including `null` and `undefined`) it is wrapped in an `Ok` variant. If the function throws an error it is wrapped in an `Err` variant. Errors that are not of type `Error` are wrapped in an `Error` object.

**Example:**

```ts
expect(fromThrowable(() => 42).unwrap()).toBe(42);
expect(fromThrowable(() => ({})).unwrap()).toEqual({});
expect(fromThrowable(() => NaN).unwrap()).toEqual(NaN);
expect(fromThrowable(() => "").unwrap()).toEqual("");

expect(fromThrowable(() => null).unwrap()).toEqual(null);
expect(fromThrowable(() => undefined).unwrap()).toEqual(undefined);

const a = fromThrowable(() => {
  throw "error";
});

expect(a.unwrapErr()).toEqual(new Error("error"));

const b = fromThrowable(() => {
  throw new Error("error");
});

expect(b.unwrapErr()).toEqual(new Error("error"));
```

### Methods

#### `isOk: boolean`

Property that is `true` if the result is Ok.

**Example:**

```ts
expect(ok(42).isOk).toBe(true);
expect(err("error").isOk).toBe(false);
```

#### `isOkAnd: (predicate: (value: T) => boolean) => boolean`

Returns `true` if the result is an Ok variant and the value inside of it matches a predicate.

**Example:**

```ts
expect(some(42).isSomeAnd((v) => v % 2)).toBe(true);
expect(some(42).isSomeAnd((v) => v > 100)).toBe(true);
```

#### `isErr: boolean`

Property that is `true` if the result is Err.

**Example:**

```ts
expect(ok(42).isErr).toBe(false);
expect(err("error").isErr).toBe(true);
```

#### `isErrAnd: (predicate: (value: E) => boolean) => boolean`

Returns `true` if the result is an Err variant and the value inside of it matches a predicate.

**Example:**

```ts
expect(some(42).isSomeAnd((v) => v % 2)).toBe(true);
expect(some(42).isSomeAnd((v) => v > 100)).toBe(true);
```

#### `ok: () => Option<T>`

Converts from Result<T, E> to Option<T> discarding the error, if any.

**Example:**

```ts
expect(ok(42).ok()).toEqual(some(42));
expect(err("error").ok()).toEqual(none);
```

#### `err: () => Option<T>`

Converts from Result<T, E> to Option<E> discarding the error, if any.

**Example:**

```ts
expect(ok(42).ok()).toEqual(some(42));
expect(err("error").ok()).toEqual(none);
```

#### `expect: (message: string) => T`

Returns the inner value. Useful for situations where if the result is an `Err` then there is no use to continue running the program. The message can be used for debugging purposes.

- `message`: The message to throw if the option is an `Err` variant.

**Throws:**

- Throws the given message if the option is an `Err` variant.

**Example:**

```ts
expect(ok(42).expect("should be a number")).toBe(42);
expect(err("error").expect("this will throw")).toThrow();
```

#### `unwrap: () => T`

Returns the Ok value or throws an Error if the result is an `Err` variant.

**Throws:**

- Throws if the option is a `None` variant.

**Example:**

```ts
expect(ok(42).unwrap()).toBe(42)
expect(err("error").unwrap().toThrow()
```

#### `expectErr: (message: string) => T`

Returns the contained Err value or throws an error if the result is an `Ok` variant.

- `message`: The message to throw if the option is an `Ok` variant.

**Throws:**

- Throws the given message if the option is an `Ok` variant.

**Example:**

```ts
expect(err(42).expectErr("should be a number")).toBe(42);
expect(ok(42).expectErr("this will throw")).toThrow();
```

#### `unwrapErr: () => E`

Returns the contained Err value or throws an error if the result is an `Ok` variant.

**Throws:**

- Throws if the option is a `None` variant.

\*\*Example

:\*\*

```ts
expect(err("error").unwrapErr()).toBe(42)
expect(ok(42).unwrapErr().toThrow()
```

#### `unwrapOr: (defaultValue: T) => T`

Returns the Ok value or the provided default value.

- `defaultValue`: The default value to return if the result is an `Err` variant.

**Example:**

```ts
expect(ok(42).unwrapOr(1)).toBe(42)
expect(err("error").unwrapOr(1).toBe(1)
```

#### `unwrapOrElse: (defaultValueFun: () => T) => T`

Returns the Ok value or computes it from a function.

- `defaultValueFun`: The function that computes the default value if the result is an `Err` variant.

**Example:**

```ts
expect(ok(42).unwrapOrElse(() => 1)).toBe(42)
expect(err("error").unwrapOrElse(() => 1).toBe(1)
```

#### `map: <U>(f: (value: T) => U) => Result<U, E>`

Maps a Result<T, E> to Result<U, E> by applying a function to a contained Ok value, leaving an Err value untouched.

**Example:**

```ts
expect(
  ok(42)
    .map((x) => x + 1)
    .unwrap()
).toBe(43);
expect(
  err("error")
    .map((x) => x + 1)
    .isNone()
).toBe(true);
```

#### `mapOr: <U>(defaultValue: U, f: (value: T) => U) => U`

Returns the provided default (if Err), or applies a function to the contained value (if Ok).

- `defaultValue`: The default value to return if the result is an `Err` variant.
- `f`: The function to apply to the contained value if the result is an `Ok` variant.

**Example:**

```ts
expect(ok(42).mapOr(10, (x) => x + 1)).toBe(43);
expect(error("error").mapOr(10, (x) => x + 1)).toBe(11);
```

#### `mapOrElse: <U>(defaultValueFun: () => U, f: (arg: T) => U) => U`

Maps a Result<T, E> to U by applying a fallback function default to a contained Err value, or function f to a contained Ok value.

- `defaultValueFun`: The fallback function to apply to the contained Err value.
- `f`: The function to apply to the contained Ok value.

**Example:**

```ts
expect(ok(42).mapOrElse(() => 10, (x) => x + 1))).toBe(43)
expect(error("error").mapOrElse(() => 10, (x) => x + 1))).toBe(11)
```

#### `mapErr: <U>(f: (value: E) => U) => Result<T, U>`

Maps a Result<T, E> to Result<T, F> by applying a function to a contained Err value, leaving an Ok value untouched.

**Example:**

```ts
expect(
  ok(42)
    .map((x) => x + 1)
    .unwrap()
).toBe(43);
expect(
  err("error")
    .map((x) => x + 1)
    .isNone()
).toBe(true);
```

#### `and: <U>(b: Result<U, E>) => Result<U, E>`

Returns `res` if the result is Ok, otherwise returns the Err value of self.

- `b`: The Result value to return if the current result is Ok.

#### `or: <F>(b: Result<T, F>) => Result<T,F>`


library/CHANGELOG.md
# this-is-ok

## 0.8.0

### Minor Changes

- 52a3393: - replace overloaded Do with Do and DoAsync
  - add `sequence` function that transforms Option<T>[] into Option<T[]> (and simile for Result)

## 0.7.0

### Minor Changes

- 6cd4143: - Added support for async Do (for both monads)
  - BREAKING: removed the .do API in favor of the standalone Do

## 0.6.2

### Patch Changes

- fix in of type in result

## 0.6.1

### Patch Changes

- fix types for err variant

## 0.6.0

### Minor Changes

- Update internal implemenation to be much more memory efficient

## 0.5.0

### Minor Changes

- 2186ea6: update types for methods such as unwrapOr, unwrapOrElse, ... so that type inference works better when T is a union type. In other words a fix for type variance.

## 0.4.3

### Patch Changes

- Do as standalone function

## 0.4.2

### Patch Changes

- fix type bug in unwrap and expect

## 0.4.1

### Patch Changes

- f1c7cd9: fix an issue with type narrowing for Option.isNone

## 0.4.0

### Minor Changes

- d5e3c01: add `tap` method to Result

### Patch Changes

- 884f96a: fix type in unwrapEr methodr
- f0d32e8: better type narrowing

## 0.3.0

### Minor Changes

- - update README to add table of contents and API reference
  - change behaviour of from function in option
  - add craetion function (of, from, fromThrowable) for Result

## 0.2.2

### Patch Changes

- add result types versions to package.json

## 0.2.1

### Patch Changes

- full test coverage
- acadf38: Added unit tests for Result and fixed a few bugs

## 0.2.0

### Minor Changes

- result type

### Patch Changes

- 733f631: fix imports with exports and typesVersions fields in package.json

## 0.1.0

### Minor Changes

- 7628b54: Option monad


library/src/option/option.ts
import { Result, err, ok } from "../result";

export type NoneVariant<T> = {
  readonly variant: "none";
  isSome: false;
  isNone: true;
} & OptionProperties<T>;

export type SomeVariant<T> = {
  readonly variant: "some";
  readonly value: T;
  isSome: true;
  isNone: false;
} & OptionProperties<T>;

export type Option<T> = NoneVariant<T> | SomeVariant<T>;

type OptionProperties<T> = {
  /**
   *  Property that is `true` if the option is a `Some` variant and `false` otherwise.
   *
   * @example
   * ```ts
   *  expect(some(3).isSome).toBe(true)
   *  expect(none.isSome).toBe(false)
   * ```
   */
  isSome: boolean;

  /**
   *  Returns true if the option is a Some variant and the value inside
   *  of it matches a predicate.
   *
   * @example
   * ```ts
   *  expect(some(42).isSomeAnd((v) => v % 2)).toBe(true)
   *  expect(some(42).isSomeAnd((v) => v > 100)).toBe(true)
   * ```
   */
  isSomeAnd: (predicate: (value: T) => boolean) => boolean;

  /**
   *  Property that is `true` if the option is a `None` variant and `false` otherwise.
   *
   * @example
   * ```ts
   *  expect(none.isNone).toBe(true)
   *  expect(some(42).isNone).toBe(false)
   * ```
   */
  isNone: boolean;

  /**
   *  Returns the inner value. Useful for situations where if the
   *  option is a `None` then there is no tap to continue running the program.
   *  The message can be used for debugging purposes.
   *
   * @param message the message to throw it the option is a `None` variant.
   * @throws throws the given message if the option is a `None` variant.
   *
   * @example
   * ```ts
   *  expect(some(42).expect("should be a number")).toBe(42)
   *  expect(none.expect("this will throw")).toThrow()
   * ```
   */
  expect: (message: string) => T;

  /**
   *  Returns the inner value or throws an Error if the option is a `None` variant.
   *
   * @throws throws if the option is a `None` variant.
   *
   * @example
   * ```ts
   *  expect(some(42).unwrap()).toBe(42)
   *  expect(none.unwrap().toThrow()
   * ```
   */
  unwrap: () => T;

  /**
   *  Returns the inner value or the provided default value.
   *
   * @example
   * ```ts
   *  expect(some(42).unwrapOr(1)).toBe(42)
   *  expect(none.unwrapOr(1).toBe(1)
   * ```
   */
  unwrapOr: <T1 extends T>(defaultValue: T1) => T;

  /**
   *  Returns the inner value or computes it from a function.
   *
   * @example
   * ```ts
   *  expect(some(42).unwrapOrElse(() => 1)).toBe(42)
   *  expect(none.unwrapOrElse(() => 1).toBe(1)
   * ```
   */
  unwrapOrElse: <T1 extends T>(defaultValueFun: () => T1) => T;

  /**
   *  Maps an Option<T> to Option<U> by applying a function to a contained
   *  value (if Some) or returns None (if None).
   *
   * @example
   * ```ts
   *  expect(some(42).map((x) => x + 1).unwrap()).toBe(43)
   *  expect(none.map((x) => x + 1).isNone()).toBe(true)
   * ```
   */
  map: <U>(f: (value: T) => U) => Option<U>;

  /**
   *  Returns the provided default result (if none),
   *  or applies a function to the contained value (if any).
   *
   * @example
   * ```ts
   *  expect(some(42).mapOr(10, (x) => x + 1).unwrap()).toBe(43)
   *  expect(none.mapOr(10, (x) => x + 1).unwrap()).toBe(11)
   * ```
   */
  mapOr: <U>(defaultValue: U, f: (value: T) => U) => U;

  /**
   *  Computes a default function result (if none),
   *  or applies a different function to the contained value (if any).
   *
   * @example
   * ```ts
   *  expect(some(42).mapOrElse(() => 10, (x) => x + 1)).unwrap()).toBe(43)
   *  expect(none.mapOrElse(() => 10, (x) => x + 1)).unwrap()).toBe(11)
   * ```
   */
  mapOrElse: <U>(defaultValueFun: () => U, f: (arg: T) => U) => U;

  okOr: <E>(err: E) => Result<T, E>;
  okOrElse: <E>(f: () => E) => Result<T, E>;

  /**
   *  Returns None if the option is None, otherwise returns `b`.
   *
   * @example
   * ```ts
   *  expect(some(42).and(some(1)).unwrap()).toBe(1)
   *  expect(none.and(some(1)).isNone).toBe(true)
   * ```
   */
  and: <U>(b: Option<U>) => Option<U>;

  /**
   *  Returns None if the option is None, otherwise returns `b`.
   *
   * @example
   * ```ts
   *  expect(some(42).or(some(1)).unwrap()).toBe(43)
   *  expect(none.or(some(1)).unwrap()).toBe(1)
   * ```
   */
  or: <T1 extends T>(b: Option<T1>) => Option<T>;

  /**
   * Returns the option if it contains a value,
   *  otherwise calls f and returns the result.
   *
   * @example
   * ```ts
   * expect(some(42).orElse(() => some(1)).unwrap()).toBe(42)
   * expect(none.orElse(() => some(1)).unwrap()).toBe(1)
   * ```
   */
  orElse: <T1 extends T>(f: () => Option<T1>) => Option<T>;

  /**
   *  Returns None if the option is None, otherwise calls
   *  f with the wrapped value and returns the result.
   *
   * @alias andThen
   *
   * @example
   * ```ts
   *  expect(some(42).flatMap(x => some(x + 1)).unwrap()).toBe(43)
   *  expect(some(42).flatMap(x => none).isNone()).toBe(true)
   * ```
   */
  flatMap: <U>(f: (value: T) => Option<U>) => Option<U>;

  /**
   *  Returns None if the option is None, otherwise calls
   *  f with the wrapped value and returns the result.
   *
   * @alias flatMap
   *
   * @example
   * ```ts
   *  expect(some(42).andThen(x => some(x + 1)).unwrap()).toBe(43)
   *  expect(some(42).andThen(x => none).isNone()).toBe(true)
   * ```
   */
  andThen: <U>(f: (value: T) => Option<U>) => Option<U>;

  /**
   *  Runs the given void function with the inner value if the option is a `Some` variant
   *  or does nothing if the option is a `None` variant.
   *
   * @alias flatMap
   *
   * @example
   * ```ts
   *  expect(some(42).andThen(x => some(x + 1)).unwrap()).toBe(43)
   *  expect(some(42).andThen(x => none).isNone()).toBe(true)
   * ```
   */
  tap: <F extends void | Promise<void>>(f: (value: T) => F) => F;

  /**
   * Returns `None` if the option is `None`,
   * otherwise returns `None` if the predicate `predicate` returns `false` when applied to the contained
   * value, otherwise returns the `Option<T>`.
   *
   * @example
   * ```ts
   * expect(some(42).filter(x => x > 10).unwrap()).toBe(42)
   * expect(some(42).filter(x => x < 10).isNone()).toBe(true)
   * ```
   */
  filter: (predicate: (arg: T) => boolean) => Option<T>;

  /**
   * 
   * @example
   * ```ts
   * expect(
      some(3).match({
        some: (d) => d + 1,
        none: () => 0,
      })
    ).toBe(4);

    expect(
      none.match({
        some: (d) => d + 1,
        none: () => 0,
      })
    ).toBe(0);
   * ```
   */
  match: <U>(pattern: { some: (value: T) => U; none: () => U }) => U;

  bind: () => T;
  b: () => T;
};

const noneHandler: ProxyHandler<Option<any>> = {
  get: (_, prop) => {
    switch (prop) {
      case "variant":
        return "none";
      case "isSome":
        return false;
      case "isNone":
        return true;
      case "isSomeAnd":
        return () => false;
      case "expect":
        return (message: any) => {
          throw new Error(`Error: ${message}`);
        };
      case "unwrap":
        return () => {
          throw new Error("Error: called `.unwrap()` on none");
        };
      case "unwrapOr":
        return <T>(defaultValue: T) => defaultValue;
      case "unwrapOrElse":
        return <T>(getDefaultValue: () => T) => getDefaultValue();
      case "map":
        return () => none;
      case "mapOr":
        return (defaultValue: any) => defaultValue;
      case "mapOrElse":
        return (defaultValue: any) => defaultValue();
      case "and":
        return () => none;
      case "flatMap":
        return () => none;
      case "andThen":
        return () => none;
      case "filter":
        return () => none;
      case "okOr":
        return (error: any) => err(error);
      case "okOrElse":
        return (error: any) => err(error());
      case "or":
        return (b: any) => b;
      case "orElse":
        return (f: any) => f();
      case "tap":
        return <F extends void | Promise<void>>(f: (_: any) => F): F => {
          return undefined as any;
        };
      case "match":
        return (pattern: any) => pattern.none();
      case "bind":
        return () => {
          throw "bind";
        };
      case "b":
        return () => {
          throw "bind";
        };
    }
  },
};

export const none: Option<any> = new Proxy(
  { variant: "none" } as Option<any>,
  noneHandler
);

const someHandler: ProxyHandler<Option<any>> = {
  get: (target, prop) => {
    const { value } = target as SomeVariant<any>;
    const flatMap = <U>(f: (value: any) => Option<U>): Option<U> => f(value);
    const unwrap = (): any => value;

    switch (prop) {
      case "variant":
        return "some";
      case "value":
        return value;
      case "isSome":
        return true;
      case "isSomeAnd":
        return (predicate: any) => predicate(value);
      case "isNone":
        return false;
      case "expect":
        return unwrap;
      case "unwrap":
        return unwrap;
      case "unwrapOr":
        return unwrap;
      case "unwrapOrElse":
        return unwrap;
      case "map":
        return (f: any) => some(f(value));
      case "mapOr":
        return (_: any, f: any) => f(value);
      case "mapOrElse":
        return (defaultValue: any, f: any) => f(value);
      case "and":
        return (b: any) => b;
      case "flatMap":
        return flatMap;
      case "andThen":
        return flatMap;
      case "filter":
        return (predicate: any) => (predicate(value) ? some(value) : none);
      case "or":
        return () => some(value);
      case "okOr":
        return () => ok(value);
      case "okOrElse":
        return () => ok(value);
      case "orElse":
        return () => some(value);
      case "tap":
        return (f: any) => f(value);
      case "match":
        return (pattern: any) => pattern.some(value);
      case "bind":
        return unwrap;
      case "b":
        return unwrap;
    }
  },
};

export const some = <T>(value: T): Option<T> =>
  new Proxy({ variant: "some", value } as Option<T>, someHandler);
